pkg/klusterlet/clusterinfo/loggingInfo_sync.go

```diff
@@ -0,0 +1,254 @@
+	var serviceType = corev1.ServiceTypeClusterIP
+	if kubeVendor != clusterv1beta1.KubeVendorOpenShift {
+		serviceType = corev1.ServiceTypeLoadBalancer
+	}
+	owner, err := s.getOwnerRef(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get owner. error: %v", err)
+	}
+	newAgentService := s.newService(serviceType, *owner)
+
+	agentService, err := s.kubeClient.CoreV1().Services(s.agentNamespace).Get(ctx, s.agentName, metav1.GetOptions{})
+	switch {
+	case errors.IsNotFound(err):
+		agentService, err = s.kubeClient.CoreV1().Services(s.agentNamespace).Create(ctx, newAgentService, metav1.CreateOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("failed to create agent service. error:%v", err)
+		}
+		break
+	case err != nil:
+		return nil, fmt.Errorf("failed to get agent service. error:%v", err)
+	case needUpdateService(agentService, newAgentService):
+		agentService, err = s.kubeClient.CoreV1().Services(s.agentNamespace).Update(ctx, newAgentService, metav1.UpdateOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("failed to create agent service. error:%v", err)
+		}
+	}
+
+	if s.clusterName == "local-cluster" {
+		endpoint := &corev1.EndpointAddress{Hostname: fmt.Sprintf("%s.%s.svc", s.agentName, s.agentNamespace)}
+		return endpoint, nil
+	}
+
+	if kubeVendor != clusterv1beta1.KubeVendorOpenShift {
+		return s.getEndpointAddressFromService(ctx)
+	}
+
+	// for OCP, get hostName from route
+	return s.getEndpointAddressFromRoute(ctx)
+}
+
+func (s *loggingInfoSyncer) getEndpointAddressFromService(ctx context.Context) (*corev1.EndpointAddress, error) {
+	klSvc, err := s.kubeClient.CoreV1().Services(s.agentNamespace).Get(ctx, s.agentName, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	if klSvc.Spec.Type != corev1.ServiceTypeLoadBalancer {
+		return nil, fmt.Errorf("agent service should be in type of loadbalancer")
+	}
+
+	if len(klSvc.Status.LoadBalancer.Ingress) == 0 {
+		return nil, fmt.Errorf("agent load balancer service does not have valid ip address")
+	}
+
+	// Only get the first IP and host name
+	endpoint := &corev1.EndpointAddress{
+		IP:       klSvc.Status.LoadBalancer.Ingress[0].IP,
+		Hostname: klSvc.Status.LoadBalancer.Ingress[0].Hostname,
+	}
+
+	return endpoint, nil
+}
+
+func (s *loggingInfoSyncer) getEndpointAddressFromRoute(ctx context.Context) (*corev1.EndpointAddress, error) {
+	endpoint := &corev1.EndpointAddress{}
+	ingressDomain, err := s.getIngressDomain(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get ingress domain.error %v", err)
+	}
+
+	owner, err := s.getOwnerRef(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get owner. error: %v", err)
+	}
+	newRoute := s.newRoute(*owner)
+	route, err := s.routeV1Client.RouteV1().Routes(s.agentNamespace).Get(ctx, s.agentName, metav1.GetOptions{})
+	switch {
+	case errors.IsNotFound(err):
+		endpoint.Hostname = fmt.Sprintf("%s-%s.%s", s.agentName, rand.String(6), ingressDomain)
```

Comment:
will always use the random string? or:
- use name-namespace if its length < 63
- name-hash(namespace)

Reply:
I think using name-hash may be more simple here.